File: apps/agni-server/src/core/interactions/patrimony/updatePatrimony.ts
import { IUsecase } from "../interfaces";
import { ResourceNotFoundError } from "@core/errors/resournceNotFoundError";
import { mapperPatrimonyType } from "@core/domains/constants";
import Repository from "@core/adapters/repository";
import { Patrimony } from "@core/domains/entities/patrimony";
import { Account } from "@core/domains/entities/account";

export type RequestUpdatePatrimony = {
    patrimonyId: string
    title?: string
    amount?: number
    accountIds?: string[]
    type?: string
}

export class UpdatePatrimonyUseCase implements IUsecase<RequestUpdatePatrimony, void> {
    private patrimonyRepo: Repository<Patrimony>
    private accountRepo: Repository<Account>

    constructor(patrimonyRepo: Repository<Patrimony>, accountRepo: Repository<Account>) {
        this.patrimonyRepo = patrimonyRepo
        this.accountRepo = accountRepo
    }

    async execute(request: RequestUpdatePatrimony): Promise<void> {
        const patrimony = await this.patrimonyRepo.get(request.patrimonyId)

        if (!patrimony)
            throw new ResourceNotFoundError("PATRIMONY_NOT_FOUND")

        if (request.title)
            patrimony.setTitle(request.title)

        if (request.amount)
            patrimony.setAmount(request.amount)

        if (request.accountIds) {
            if ((await this.accountRepo.getManyByIds(request.accountIds)).length === 0)
                throw new ResourceNotFoundError("SOME_ACCOUNT_NOT_FOUND") 
        }

        if (request.type)
            patrimony.setType(mapperPatrimonyType(request.type)) 

        if (patrimony.hasChange())
            await this.patrimonyRepo.update(patrimony)
    }
}

File: apps/agni_api/src/main/kotlin/dev/auguste/agni_api/core/usecases/patrimonies/UpdatePatrimony.kt
package dev.auguste.agni_api.core.usecases.patrimonies

import dev.auguste.agni_api.core.adapters.repositories.IRepository
import dev.auguste.agni_api.core.entities.Account
import dev.auguste.agni_api.core.entities.Patrimony
import dev.auguste.agni_api.core.usecases.interfaces.IUseCase
import dev.auguste.agni_api.core.usecases.patrimonies.dto.UpdatePatrimonyInput

class UpdatePatrimony(
    private val patrimonyRepo: IRepository<Patrimony>,
    private val accountRepo: IRepository<Account>): IUseCase<UpdatePatrimonyInput, Unit> {
    override fun execAsync(input: UpdatePatrimonyInput) {
        val patrimony = patrimonyRepo.get(input.id) ?: throw Error("Patrimony ${input.id} not found")

        if (input.title != null)
            patrimony.title = input.title

        if (input.amount != null)
            patrimony.amount = input.amount

        if (input.accountIds != null) {
            if (input.accountIds.isEmpty())
                if (this.accountRepo.getManyByIds(input.accountIds).size != input.accountIds.size)
                    throw Error("Not all accounts are found")

            patrimony.accountIds = input.accountIds.toMutableSet()
        }

        if (input.type != null) {
            patrimony.type = input.type
        }

        if (patrimony.hasChanged())
            patrimonyRepo.update(patrimony)
    }
}